ui/appui-react/src/appui-react/ui-items-provider/AbstractUiItemsManager.ts

```diff
@@ -0,0 +1,548 @@
+
+  if (isAbstractStatusBarCustomItem(abstractItem)) {
+    const content: React.ReactNode | undefined = (abstractItem as any).reactNode;
+    return {
+      ...abstractItem,
+      badge: abstractItem.badgeType,
+      content,
+    };
+  } else {
+    const icon = fromAbstractIcon(abstractItem.icon, abstractItem.internalData);
+    item = {
+      ...abstractItem,
+      badge: abstractItem.badgeType,
+      icon,
+    };
+  }
+  assert(!!item);
+  setOriginalData(item, abstractItem);
+  return item;
+}
+
+function fromAbstractToolbarItem(abstractItem: AbstractToolbarItem): ToolbarItem {
+  let item = getOriginalData<ToolbarItem>(abstractItem);
+  if (item)
+    return item;
+
+  const icon = fromAbstractIcon(abstractItem.icon, abstractItem.internalData);
+  if (AbstractToolbarItemUtilities.isCustomDefinition(abstractItem)) {
+    item = {
+      ...abstractItem,
+      badge: abstractItem.badgeType,
+      icon,
+    };
+  } else {
+    item = {
+      ...abstractItem,
+      badge: abstractItem.badgeType,
+      parentGroupItemId: abstractItem.parentToolGroupId,
+      icon,
+    };
+  }
+  setOriginalData(item, abstractItem);
+  return item;
+}
+
+function fromAbstractWidget(abstractWidget: AbstractWidget): Widget {
+  let widget = getOriginalData<Widget>(abstractWidget);
+  if (widget)
+    return widget;
+
+  const allowedPanels = abstractWidget.allowedPanelTargets
+    // @ts-ignore Possibly 'any'
+    ?.map((target) => {
+      const map = {
+        top: StagePanelLocation.Top,
+        left: StagePanelLocation.Left,
+        right: StagePanelLocation.Right,
+        bottom: StagePanelLocation.Bottom,
+      };
+      // @ts-ignore Possibly 'any'
+      return map[target];
+    });
+  const icon = fromAbstractIcon(abstractWidget.icon, abstractWidget.internalData);
+  widget = {
+    ...abstractWidget,
+    id: abstractWidget.id ?? "",
+    content: abstractWidget.getWidgetContent(),
+    allowedPanels,
+    badge: abstractWidget.badgeType,
+    canFloat: abstractWidget.isFloatingStateSupported && {
+      containerId: abstractWidget.floatingContainerId,
+      defaultPosition: abstractWidget.defaultFloatingPosition,
+      defaultSize: abstractWidget.defaultFloatingSize,
+      hideWithUi: abstractWidget.hideWithUiWhenFloating,
+      isResizable: abstractWidget.isFloatingStateWindowResizable,
+    },
+    icon,
+  };
+  setOriginalData(widget, abstractWidget);
+  return widget;
+}
+
+function fromAbstractStagePanelLocation(location: AbstractStagePanelLocation): StagePanelLocation {
+  switch (location) {
+    case AbstractStagePanelLocation.Left:
+      return StagePanelLocation.Left;
+    case AbstractStagePanelLocation.Top:
+      return StagePanelLocation.Top;
+    case AbstractStagePanelLocation.TopMost:
+      return StagePanelLocation.Top;
+    case AbstractStagePanelLocation.Bottom:
+      return StagePanelLocation.Bottom;
+    case AbstractStagePanelLocation.BottomMost:
+      return StagePanelLocation.Bottom;
```

Comment:
`fromAbstractStagePanelLocation` is only used when requesting widgets to an adapted UiItemsProvider, it should not return anything for TopMost and BottomMost because if every section is requested, the widget will return in both places. (This is linked to my other comment about the adapted UiItemsManager...)

Same reflexion goes to `fromAbstractStagePanelSection` below.

Reply:
In that case it would not be possible to query widgets from appui-react provider that were registered as TopMost via abstract provider? We should be handling duplicate widgets and I believe it's not a huge problem to return the same widgets, similar to when the widget provider ignores one of the arguments..

Reply:
appui-react provider would not be answering to TopMost as it is not in the Enum they are built for. But you are right that we are probably already handling duplicates, so that is probably not a problem.